ldap/servers/slapd/daemon.c

```diff
@@ -1505,6 +1573,418 @@ daemon_register_reslimits(void)
+
+    lt_epoll_events = (struct epoll_event *)slapi_ch_calloc(listen_table_size, sizeof(*lt_epoll_events));
+
+    epoll_arm_listen_pds(the_listen_table, lt_epoll_fd, EPOLL_PD_ARM);
+
+    while (!g_get_shutdown()) {
+        /* Do we need to accept new connections? */
+        int accept_new_connections = ((ct->size - g_get_current_conn_count()) > slapdFrontendConfig->reservedescriptors);
+        last_accept_new_connections = accept_new_connections;
+        if (!accept_new_connections) {
+            if (last_accept_new_connections) {
+                slapi_log_err(SLAPI_LOG_ERR, "epoll_accept_thread",
+                              "Not listening for new connections - too many fds open\n");
+            }
+            /* Need a sleep delay here. */
+            PR_Sleep(pr_timeout);
+            continue;
+        } else {
+            /* Log that we are now listening again */
+            if (!last_accept_new_connections && last_accept_new_connections != -1) {
+                slapi_log_err(SLAPI_LOG_ERR, "epoll_accept_thread",
+                              "Listening for new connections again\n");
+            }
+        }
+        epoll_fds = epoll_wait(lt_epoll_fd, lt_epoll_events, listeners, slapd_wakeup_timer);
+        switch (epoll_fds) {
+        case 0: /* Timeout */
+            break;
+        case -1: /* Error */
+            prerr = PR_GetError();
+            slapi_log_err(SLAPI_LOG_TRACE, "epoll_accept_thread", "PR_Poll() failed, " SLAPI_COMPONENT_NAME_NSPR " error %d (%s)\n",
+                          prerr, slapd_system_strerror(prerr));
+            break;
+        default: /* a new connection */
+            for (i = 0; i < epoll_fds; i++) {
+                c = lt_epoll_events[i].data.ptr;
+                handle_epoll_listener(c);
+            }
+            break;
+        }
+    }
+
+    epoll_arm_listen_pds(the_listen_table, lt_epoll_fd, EPOLL_PD_DISARM);
+    free_epoll_listen_table(the_listen_table);
+
+    /* free the epoll event array */
+    slapi_ch_free((void **)&lt_epoll_events);
+
+    /* free the listener indexes */
+    slapd_sockets_ports_free(ports);
+}
+
+static void
+epoll_pr_idle_pds(Connection_Table *ct)
+{
+    Connection *c = NULL;
+    Connection *next = NULL;
+    time_t curtime = slapi_current_rel_time_t();
+
+    /* Walk down the list of active connections to find idle connections. */
+    c = connection_table_get_first_active_connection(ct);
+    while (c) {
+        next = connection_table_get_next_active_connection(ct, c);
+        if (c->c_state == CONN_STATE_FREE) {
+            epoll_arm_pd(c, ct->epollfd, EPOLL_PD_DISARM);
+            connection_table_move_connection_out_of_active_list(ct, c);
+        } else {
+            /* we try to acquire the connection mutex, if it is already
+             * acquired by another thread, don't wait
+             */
+            if (pthread_mutex_trylock(&(c->c_mutex)) == EBUSY) {
```

Comment:
Could consider doing rc = trylock, then check rc != 0, and if it's a weird code log that (weird being not EBUSY, see man 3 pthread_mutex_trylock )

Reply:
Updates per suggestion, please review new update.